notes-0319-inductive-types.tex

```diff
@@ -0,0 +1,228 @@
+Then if this path were identified with \textit{refl}, we would have point-wise identifications between the identity and negation functions, and in particular an identification between their images at $0$, namely an element of type $Id_2(0,1)$.
+But in homework $3$ we showed that from this we can derive an element of $\bot$, so we have a contradiction.
+So $U$ can't be a set.
+\end{proof}
+
+\section{Inductive Type}
+We define an inductive type to be a type that is freely generated by a finite set of constructors, where intuitively we think of ``freely generated'' to mean that the elements of the type are all the objects that can be obtained by applying and composing the constructors. It also means that we don't impose any more relations on elements than we are absolutely required to. If you are familiar with freely generated groups from algebra then this should feel very similar.
+
+\begin{example}
+  $\2$ is generated by $0 : \2$ and $1 : \2$. Since neither constructor is a function type, there are no operations we can perform on these constructors. Therefore the type freely generated by these constructors has only two elements, as we would expect.
+\end{example}
+
+\begin{example}
+  $A + B$ is generated by $inl : A \to A + B$ and $inr : B \to A + B$. In this case the constructors are functions, so we get some more interesting elements of the inductive type. The type freely generated by these constructors contains elements like $inl\ a$ for elements $a : A$. 
+\end{example}
+    
+The first more interesting case of an inductive type is $\N$, which is generated by $zero : \N$ and $suc: \N \to \N$. This illustrates the fact that the type being defined can appear in both the domain and the co-domain of a constructor. The $suc$ constructor represents the construction of $n+1$ from a construction of $n$. Intuitively, the type freely generated by these constructors should contain $zero$, $suc(zero)$, $suc(suc(zero))$, and so on, since these are exactly the things we can construct from these constructors using the operations we have available. We can formalize this intuition with the following rules for the type $\N$: 
+
+\begin{multicols}{2}
+  \noindent
+  \begin{equation*}
+    \begin{prooftree}
+      \infer0[$0$-Introduction]{\oftype{zero}{\N}}
+    \end{prooftree}
+  \end{equation*}
+  \begin{equation*}
+    \begin{prooftree}
+      \hypo{\oftype{M}{\N}}
+      \infer1[$suc$-Introduction]{\oftype{suc\ M}{\N}}
+    \end{prooftree}
+  \end{equation*}
+\end{multicols}
+
+
+Then we would like the elimination rule to model the classic induction principle for the natural numbers. Recall that traditionally the induction principle for the set of natural numbers states that in order to prove some proposition $P(n)$ for all $n \in \N$ it suffices to prove $P(0)$ and $P(n) \rightarrow P(n+1)$. We can model this rule with the following axiom for our inductive $\N$ type:
+
+\begin{prooftree*}
+  \hypo{x : \N : \rightarrow P : U}
```

Comment:
Do ": \rightarrow" and "\rightarrow" mean the turnstile we used in lectures?

There are many sensible choices, but these two might be rare.